### PR TITLE
Improve `watchers_model.yml` and `work_packages_model.yml`

### PR DESCRIPTION
**Pull request by gohrner**
Open question: `work_packages_by_project_model.yml` is structurally
identical to `work_packages_model.yml`, the only difference is the
embedded `example:`.

I can imagine that it's important, due to the focus on API doc
generation from this spec, so one option would be to just reference the
`work_packages_model.yml` for the structural definition, avoiding
redundancy for this, but in this case leaving the redundancy in the
`example:` definition, with all possible drawbacks (inconsistencies
etc.)

Another approach would be to get rid of
`work_packages_by_project_model.yml` completely and just using the
generic `work_packages_model.yml` everywhere where we have a list of
Work Packages.

Please show some patience, as I'm not yet very experienced with OpenAPI V3.

See also: https://community.openproject.org/topics/14323

### CODE REVIEW

File: docs/api/apiv3/components/schemas/watchers_model.yml [hunk 1]
```diff
@@ -1,6 +1,43 @@
 # Schema: WatchersModel
 ---
 type: object
+properties:
```
Comment (Kharonus) [inline docs/api/apiv3/components/schemas/watchers_model.yml]: I just saw this and made my head around how we can avoid redefining collection properties again and again. So, the solution is using this schema:

```
allOf:
  - "$ref": "./collection_model.yml"
  - type: object
    required:
      - _links
      - _embedded
    properties:
      _links:
        type: object
        required:
          - self
        properties:
          self:
            allOf:
              - "$ref": "./link.yml"
              - description: |-
                  The watcher list
                  
                  **Resource**: WatchersModel
                readOnly: true
      _embedded:
        type: object
        properties:
          elements:
            type: array
            readOnly: true
            items:
              "$ref": "./user_model.yml"
```

you can start using it here, whenever I continue improving the spec, I'll use the same pattern. And please check, if this affects your java client generation.

please note, that I removed the `allOf` in the `items` section of your proposal. This is due to the fact, that it is not needed. The description at this level should refer to a single work package and we better should provide a description within the `work_package_model.yml`

File: docs/api/apiv3/components/schemas/work_packages_model.yml [hunk 2]
```diff
@@ -1,6 +1,43 @@
 # Schema: Work_PackagesModel
 ---
 type: object
+properties:
+  total:
```
Comment (Kharonus) [inline docs/api/apiv3/components/schemas/work_packages_model.yml]: please use the same pattern here as described in previous comment

File: docs/api/apiv3/components/schemas/work_packages_by_project_model.yml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # Schema: Work_Packages_by_ProjectModel
 ---
-type: object
+$ref: "./work_packages_model.yml"
```
Comment (Kharonus) [inline docs/api/apiv3/components/schemas/work_packages_by_project_model.yml]: The response of both endpoints is created by the same service, just the scope differs. Hence, I suggest the following approach:

- remove schema `work_packages_by_project`
- reference `work_packages_model` directly from `project_work_packages.yml`

**Review by Kharonus — CHANGES_REQUESTED** [review]
We need to adjust the changes a bit, to come up with a better API spec. Please check, if the proposed changes are valid for java client generation.

File: docs/api/apiv3/components/schemas/work_packages_by_project_model.yml (hunk 3 above)
Comment (gohrner) [reply docs/api/apiv3/components/schemas/work_packages_by_project_model.yml]: That's fine with me. That's actually what I had in mind first, but I then thought that maybe you'd insist on having the different examples, as the `self` link differs:

`href: "/api/v3/work_packages"` vs. `href: "/api/v3/projects/14/work_packages"`

However, the rest is just repeated verbatim...

So yes, will remove the redundant models in these cases alltogether. Note that the `self` link then however will probably be not 100% as returned by OpenProject (not tested by myself so far, though).

The PR https://github.com/opf/openproject/pull/10178 goes exactly into this direction, so it's probably (mostly) fine?

We also seem to have the same massive repetition for `available_projects_for_XXX_model.yml`, and e.g.

At first I also thought that `available_assignees_model.yml`, `available_responsibles_model.yml`, `available_watchers_model.yml` are all just `users_model.yml`s, but this does not seem to be entirely true - while the schemas are not yet properly modelled, the examples differ in the `update` or `showUser` methods. Probably intentional and thus cannot be merged with `users_model.yml`?

File: docs/api/apiv3/components/schemas/watchers_model.yml (hunk 1 above)
Comment (gohrner) [reply docs/api/apiv3/components/schemas/watchers_model.yml]: Technically seems to work, but causes the `WorkPackagesModelLinks` class to be called `WorkPackagesModelAllOfLinks`, which is not really appropriate / meaningful.

Also causes generation of addition classes `WorkPackagesModelAllOfEmbedded` which haven't been there before.

I'll have to take some more time to check if this makes sense, and maybe how to influence name generation.

In general I agree that your approach reduces redundancy in the schema definitions.

File: docs/api/apiv3/components/schemas/work_packages_by_project_model.yml (hunk 3 above)
Comment (Kharonus) [reply docs/api/apiv3/components/schemas/work_packages_by_project_model.yml]: Point about the self link is true, but it shouldn't be a different schema though. Just the example differs. In the schema, both have a `self` link, which has only an `href`.

Comment (Kharonus) [reply docs/api/apiv3/components/schemas/work_packages_by_project_model.yml]: #10178 does the same for attachments, I didn't check those yet in production, but will do now.

File: docs/api/apiv3/components/schemas/watchers_model.yml (hunk 1 above)
Comment (Kharonus) [reply docs/api/apiv3/components/schemas/watchers_model.yml]: Can you perchance provide me a glance into those generated classes? I admit, I'm kind of confused, why Java should do it this way, as it could simply create a class `WorkPackagesModel`, that has all the merged properties of `Link` and the ones mentioned there. No idea, why taking the `allOf` into the name. Or why creating an `embedded`, at all.

Comment (gohrner) [reply docs/api/apiv3/components/schemas/watchers_model.yml]: Concerning the `WatchersModel` it does exactly what you expected - this class is generated and contains the Collection Model attributes plus `_link` and `_embedded`.

I also think I understand now what happens here, but didn't have time so far to think about an alternative.

For embedded object definitions like `_link` or `_embedded` in the above example, additional classes representing these types are generated. These classes get the name `<ContainerClassName><CleanedAttributeName>` - so e.g. (just examples, these classes do NOT exists here in this case, see below) `WatchersModelLinks` or `WatchersModelEmbedded`. So far, so logical.

(Side node: My last analysis was on a wrong track concerning the generated "*Embedded" class - I should have taken a second look before replying. The class is justified, represents the object type defined for the `_embedded` attribute, and was only newly generated after your changes suggested above because before, the type defined for the `_embedded` attribute perfectly matched the type structure for the type of the "Available Assignees Model"'s `_embedded` attribute - and the generator recognized that and re-used the latter inside the `WatchersModel`. After I applied your suggested changes to "Watchers Model", this was no longer the case, and the additional `*Embedded` type for the Watcher's `_embedded` attribute had to be generated.)

Where it get's somewhat strange now is the (anonymous) object definition directly nested into the `allOf:` in your suggestion.

This causes the OpenAPI CodeGen to **also** (besides the expected `WatchersModel` class) generate an `WatchersModelAllOf` class, which only contains a `_links` and an `_embedded` attribute.

This class is then actually used / referenced from nowhere. What seems to happen here is that the same logic as for object types that are defined for attributes (like `_embedded` in the above example) is used, and instead of a property name, which is not available in this case, the `allOf` is used to generate the type suffix.

Compare the following snippets:

```yml
_links:
  type: object
```

This causes a class `<ContainerTypeName>Links` to be generated for the object definition specified in the `_links` attribute.

```yml
allOf:
  - "$ref": "./collection_model.yml"
  - type: object
```

This then (strangely?) causes a class `<ContainerTypeName>AllOf` to be generated - looks analogous to the name generation in the case before, except in the case of an attribute the type definition is nested in, this makes sense, while less so if it's nested in an `allOf` context...

However I still don't have sufficient insight into OpenAPI V3 in order to determine if this makes sense somehow, or if it's a bug in OpenAPI CodeGen.

Then it gets logical again - `WatchersModelAllOf` is the direct container type of the `_links` and `_embedded` attributes, so the types generated for these are consequentually names `WatchersModelAllOfLinks` and `WatchersModelAllOfEmbedded` - which are somewhat strange names, unfortunately.

These two types are then also used for the `links` and `embedded` attributes in the generated `WatchersModel` (without `AllOf` suffix) class.

Besides that, the generated code is as expected, though...


Two addititional questions, though:

1. In the example, methods like `showUser`, `lock` etc. show up, which are however not yet modelled properly in the spec. These will probably have to be added, the question just is if it should be done in a separate step, or in one go while adding the schema body.
2. `users_model.yml` is very similar to `watchers_model.yml` - however, in the example, the `watchers_model.yml` is missing the `update` method, which kind of make sense - so I'd understand that it may be provided by `users_model.yml`, but not `watchers_model.yml`, and thus the two need different schema definitions.
    So the remaining question here is - is it correct that the models differ in this action, and we need two models, or was it an oversight in the example, and it would make sense to unify both models into one?
    If it's an intentional difference and we would need two models, could it make sense to derive one of these two models from the other, like your approach with the `collection_model.yml`?
    So basically having an `referenced_users_model.yml` which supports `showUser`, `lock` and `delete` (and which would replace `watchers_model.yml` and `available_assignees_model.yml`, besides maybe others) while there's also an `managed_users_model.yml` which "derives" from `referenced_users_model.yml` with the `allOf` approach and adds a single method `update`? (The names are just placeholders, finding appropriate names for types / classes isn't really one of my strengths... ;)   )

File: docs/api/apiv3/components/schemas/work_packages_model.yml (hunk 2 above)
Comment (gohrner) [reply docs/api/apiv3/components/schemas/work_packages_model.yml]: Looks as if I was too slow and you already implemented / adjusted that a few days after this review.

So I'm resolving this comment.

File: docs/api/apiv3/components/schemas/work_packages_by_project_model.yml (hunk 3 above)
Comment (gohrner) [reply docs/api/apiv3/components/schemas/work_packages_by_project_model.yml]: Please have a look at the recent commit.

File: docs/api/apiv3/components/schemas/watchers_model.yml (hunk 1 above)
Comment (Kharonus) [reply docs/api/apiv3/components/schemas/watchers_model.yml]: I did generate some code myself, namely go, ts and java, and came to the same conclusion you id. every anonymous, nested object gets its own class. And due to the lack of a file path, it basically uses the fragments of the yml definition to name it. Hence, we get the `allOf` as part of the class name.

But indeed, those models are not really needed. When using the `allOf` object, the generated code correctly provides a class with all merged properties, which is the one we would want to use. I would see all unneeded classes as code garbage, but this is to be expected when using code generation. It is meant to be generic, and not use case specific.

to the additional questions:
1. I'm completely fine with modelling a partial model first and refine it at any later point in time. The API spec has holes in it, and we know it. Even modeled schemes are not expected to be always complete. If such a hole is detected, it should get fixed. But no need to force everything into a very first schema. Even an incomplete one is already an improvement.

I'm now checking question 2 and the new commit in general. Answer pending...

Comment (Kharonus) [reply docs/api/apiv3/components/schemas/watchers_model.yml]: So, about question 2:

Here both examples are misleading and wrong currently (those in `users_model.yml` and those in `watchers_model.yml`). Your premise that the watchers endpoint actually returns a collection of users, is correct. I would like to update those examples in the collection models on top of your PR, so when it is merged, I will create another one to make some polishing to `user_model.yml`, which is mostly about some missing or wrongly typed properties. and amend the examples.

File: docs/api/apiv3/components/schemas/watchers_model.yml [hunk 4]
```diff
@@ -1,6 +1,36 @@
...
+        type: object
+        properties:
+          elements:
+            type: array
+            readOnly: true
+            items:
+                allOf:
+                - "$ref": "./user_model.yml"
+                - description: Collection of Users
+                readOnly: true
```
Comment (Kharonus) [inline docs/api/apiv3/components/schemas/watchers_model.yml]: `readOnly` is misplaced here. From OpenAPI spec for v3.0.3: `Relevant only for Schema "properties" definitions.`

I'd say, remove it from here. Only thing you need is `readOnly` in line 28, which already states, that the whole array `elements` should only be validated for response bodies, and no request bodies.

**Review by Kharonus — CHANGES_REQUESTED** [review]
Only one small fix and this one is good to go for me.

File: docs/api/apiv3/components/schemas/watchers_model.yml (hunk 1 above)
Comment (gohrner) [reply docs/api/apiv3/components/schemas/watchers_model.yml]: > But indeed, those models are not really needed. When using the allOf object, the generated code correctly provides a class with all merged properties, which is the one we would want to use. I would see all unneeded classes as code garbage, but this is to be expected when using code generation. It is meant to be generic, and not use case specific.

The unused `WatchersModelAllOf` doesn't hurt at all - it's just the contained (and used) `WatchersModelAllOfLinks` and `WatchersModelAllOfEmbedded` which have somewhat peculiar names for what they do. Also doesn't really hurt, just looks a bit strange in code which uses them.

Concerning question 2 / `users_model.yml` / `watchers_model.yml`:

So did I get you right; in this PR, I'll leave them as it is, and you (or me) will unify them as a subsequent step in a new PR?> But indeed, those models are not really needed. When using the allOf object, the generated code correctly provides a class with all merged properties, which is the one we would want to use. I would see all unneeded classes as code garbage, but this is to be expected when using code generation. It is meant to be generic, and not use case specific.

The unused `WatchersModelAllOf` doesn't hurt at all - it's just the contained (and used) `WatchersModelAllOfLinks` and `WatchersModelAllOfEmbedded` which have somewhat peculiar names for what they do. Also doesn't really hurt, just looks a bit strange in code which uses them.

Concerning question 2 / `users_model.yml` / `watchers_model.yml`:

So did I get you right; in this PR, I'll leave them as it is, and you (or me) will unify them as a subsequent step in a new PR?

Comment (gohrner) [reply docs/api/apiv3/components/schemas/watchers_model.yml]: > But indeed, those models are not really needed. When using the allOf object, the generated code correctly provides a class with all merged properties, which is the one we would want to use. I would see all unneeded classes as code garbage, but this is to be expected when using code generation. It is meant to be generic, and not use case specific.

The unused `WatchersModelAllOf` doesn't hurt at all - it's just the contained (and used) `WatchersModelAllOfLinks` and `WatchersModelAllOfEmbedded` which have somewhat peculiar names for what they do. Also doesn't really hurt, just looks a bit strange in code which uses them.

Concerning question 2 / `users_model.yml` / `watchers_model.yml`:

So did I get you right; in this PR, I'll leave them as it is, and you (or me) will unify them as a subsequent step in a new PR?

Comment (gohrner) [reply docs/api/apiv3/components/schemas/watchers_model.yml]: Sorry for the double-post -
1. first I could not add the first of these two coments (tried three times),
2. then I could not edit it to remove the doublicate content which crept in during the submission retries (the content does not change after editing)
3. and now I cannot delete it (nothing happens if I select "delete" from its menu)...

Strange.

Comment (Kharonus) [reply docs/api/apiv3/components/schemas/watchers_model.yml]: Uhm, I guess there is a bit of duplication? :D

Yes, please do not merge both collection models. There IS a difference in the collection models, in the `_links` they show (on the collection, not the embedded user elements). The user elements are exactly the same in both, even use the same representer code. Hence, we can, like you have it already, `$ref` the `user_model.yml` in the `watchers_model.yml`.

edith says: At least we have some funny history in here now ;)

File: docs/api/apiv3/components/schemas/watchers_model.yml (hunk 4 above)
Comment (gohrner) [reply docs/api/apiv3/components/schemas/watchers_model.yml]: Adjusted.

File: docs/api/apiv3/components/schemas/watchers_model.yml (hunk 1 above)
Comment (gohrner) [reply docs/api/apiv3/components/schemas/watchers_model.yml]: > edith says: At least we have some funny history in here now ;)

I feel innocent... ;)

**Review by Kharonus — APPROVED** [review]
LGTM
